File: northwind-angular/src/app/services/api.service.ts
/**
 * API SERVICE — Angular Concept: Services + Dependency Injection (DI)
 *
 * React equivalent: In React, you'd use axios directly in components or create
 * a custom hook (useApi). Angular uses a different pattern:
 *
 * 1. @Injectable() — Decorator that marks this class as available for DI
 * 2. providedIn: 'root' — Registers as a singleton (one instance shared app-wide)
 * 3. HttpClient — Angular's built-in HTTP library (replaces axios)
 * 4. Observable — RxJS streams (Angular uses Observables instead of Promises)
 *
 * In Angular, services handle business logic and API calls.
 * Components inject services via constructor parameters.
 * This separation of concerns is a core Angular pattern.
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  // providedIn: 'root' means Angular creates a single instance (singleton)
  // and makes it available everywhere — no need to add it to a providers array.
  // React equivalent: A module-level singleton or React Context.
  providedIn: 'root'
})
export class ApiService {

  /**
   * Constructor Injection — Angular Concept: Dependency Injection
   *
   * Angular automatically provides an HttpClient instance when this service
   * is created. You just declare it as a constructor parameter.
   *
   * React equivalent: There's no direct equivalent. In React, you'd import
   * axios at the top of the file. Angular's DI makes testing easier because
   * you can swap HttpClient with a mock in tests.
   */
  constructor(private http: HttpClient) {}

  /**
   * Generic request method that handles all HTTP methods.
   *
   * Returns an Observable<any> — Angular Concept: Observables (RxJS)
   *
   * React equivalent: This would return a Promise (from axios).
   * Observables are like Promises but more powerful:
   * - Can emit multiple values over time (Promise resolves once)
   * - Can be cancelled (Promises can't)
   * - Have operators like map, filter, retry, debounce
   *
   * In this app, we use .subscribe() to handle the response,
   * which is similar to .then() on a Promise.
   */
  request(method: string, endpoint: string, body?: unknown): Observable<any> {
    switch (method.toUpperCase()) {
      case 'POST':
        return this.http.post(endpoint, body);
      case 'PUT':
        return this.http.put(endpoint, body);
      case 'DELETE':
        return this.http.delete(endpoint);
      default:
        return this.http.get(endpoint);
    }
  }
}

File: northwind-blazor/Services/ApiService.cs
/**
 * API SERVICE — Blazor Concept: Injectable Service with HttpClient
 *
 * React equivalent: Direct axios calls or a custom api.ts module
 * Angular equivalent: api.service.ts with @Injectable and HttpClient
 *
 * Key Blazor concepts:
 * 1. Service class registered with DI in Program.cs (AddScoped<ApiService>)
 * 2. Constructor injection of HttpClient (like Angular's constructor DI)
 * 3. async/await with Task<T> — like React's async/await with Promises
 * 4. Returns Task (not Observable like Angular) — closer to React's pattern!
 *
 * Comparison:
 *   React:   const res = await axios.get('/api/...'); return res.data;
 *   Angular: this.http.get('/api/...').subscribe(data => ...)
 *   Blazor:  var json = await _http.GetStringAsync('/api/...');
 *
 * Blazor's async/await is MORE similar to React than Angular.
 * No Observables, no subscribe() — just familiar async/await.
 */
using System.Net.Http.Json;
using System.Text.Json;

namespace NorthwindBlazor.Services;

public class ApiService
{
    private readonly HttpClient _http;

    // Constructor injection — Blazor DI provides HttpClient automatically
    // React equivalent: No DI, just "import axios from 'axios'"
    // Angular equivalent: constructor(private http: HttpClient) {}
    public ApiService(HttpClient http)
    {
        _http = http;
    }

    /// <summary>
    /// Generic request method handling all HTTP methods.
    /// Returns JsonElement for flexible dynamic JSON handling.
    ///
    /// React equivalent: axios({ method, url, data })
    /// Angular equivalent: this.http[method](endpoint, body)
    /// </summary>
    public async Task<JsonElement> RequestAsync(string method, string endpoint, object? body = null)
    {
        // C# switch expression — like a compact switch statement
        // React/JS equivalent: const response = method === 'POST' ? await axios.post(...) : ...
        HttpResponseMessage response = method.ToUpper() switch
        {
            "POST" => await _http.PostAsJsonAsync(endpoint, body),
            "PUT" => await _http.PutAsJsonAsync(endpoint, body),
            "DELETE" => await _http.DeleteAsync(endpoint),
            _ => await _http.GetAsync(endpoint),  // _ is the default case (like "default:" in switch)
        };

        response.EnsureSuccessStatusCode();
        var json = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<JsonElement>(json);
    }
}
